grpclog/loggerv2.go

```diff
@@ -167,70 +190,89 @@ func newLoggerV2() LoggerV2 {
 	})
 }
 
-func (g *loggerT) output(severity int, s string) {
+func (g *loggerT) printFunc(severity int, fn func() string) {
+	lg := g.m[severity]
+	if lg.Writer() == io.Discard {
+		return
+	}
+
 	sevStr := severityName[severity]
+
 	if !g.jsonFormat {
-		g.m[severity].Output(2, fmt.Sprintf("%v: %v", sevStr, s))
+		lg.Output(2, sevStr+": "+fn())
 		return
 	}
+
 	// TODO: we can also include the logging component, but that needs more
 	// (API) changes.
 	b, _ := json.Marshal(map[string]string{
 		"severity": sevStr,
-		"message":  s,
+		"message":  fn(),
 	})
-	g.m[severity].Output(2, string(b))
+	lg.Output(2, string(b))
+}
+
+func (g *loggerT) printf(severity int, format string, args ...any) {
+	g.printFunc(severity, func() string { return fmt.Sprintf(format, args...) })
+}
+
+func (g *loggerT) print(severity int, v ...any) {
+	g.printFunc(severity, func() string { return fmt.Sprint(v...) })
+}
+
+func (g *loggerT) println(severity int, v ...any) {
+	g.printFunc(severity, func() string { return fmt.Sprintln(v...) })
 }
 
 func (g *loggerT) Info(args ...any) {
-	g.output(infoLog, fmt.Sprint(args...))
+	g.print(infoLog, args...)
 }
 
 func (g *loggerT) Infoln(args ...any) {
-	g.output(infoLog, fmt.Sprintln(args...))
+	g.println(infoLog, args...)
 }
 
 func (g *loggerT) Infof(format string, args ...any) {
-	g.output(infoLog, fmt.Sprintf(format, args...))
+	g.printf(infoLog, format, args...)
 }
 
 func (g *loggerT) Warning(args ...any) {
-	g.output(warningLog, fmt.Sprint(args...))
+	g.print(warningLog, args...)
 }
 
 func (g *loggerT) Warningln(args ...any) {
-	g.output(warningLog, fmt.Sprintln(args...))
+	g.println(warningLog, args...)
 }
 
 func (g *loggerT) Warningf(format string, args ...any) {
-	g.output(warningLog, fmt.Sprintf(format, args...))
+	g.printf(warningLog, format, args...)
 }
 
 func (g *loggerT) Error(args ...any) {
-	g.output(errorLog, fmt.Sprint(args...))
+	g.print(errorLog, args...)
 }
 
 func (g *loggerT) Errorln(args ...any) {
-	g.output(errorLog, fmt.Sprintln(args...))
+	g.println(errorLog, args...)
 }
 
 func (g *loggerT) Errorf(format string, args ...any) {
-	g.output(errorLog, fmt.Sprintf(format, args...))
+	g.printf(errorLog, format, args...)
 }
 
 func (g *loggerT) Fatal(args ...any) {
-	g.output(fatalLog, fmt.Sprint(args...))
-	os.Exit(1)
+	g.print(fatalLog, args...)
+	os.Exit(FatalExitCode)
 }
 
 func (g *loggerT) Fatalln(args ...any) {
-	g.output(fatalLog, fmt.Sprintln(args...))
-	os.Exit(1)
+	g.println(fatalLog, args...)
+	os.Exit(FatalExitCode)
```

Comment:
nit: I think 1 is a standard fatal exit code so we may not need a constant

Reply:
Changed